packages/driver/src/cy/commands/actions/attachFile.ts

```diff
@@ -6,23 +6,100 @@ import $errUtils from '../../../cypress/error_utils'
 import $actionability from '../../actionability'
 import { addEventCoords, dispatch } from './trigger'
 
-const createFileList = (files: Cypress.FileReferenceObject[]): FileList => {
+/* dropzone.js relies on an experimental, nonstandard API, webkitGetAsEntry().
+ * https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry
+ * The behavior here varies between browsers, and though unpleasant, we have to attempt
+ * to replicate it as far as possible.
+ *
+ * In Firefox, attempting to set webkitGetAsEntry() fails - but this is fine, because
+ * Firefox also returns a useful value from item.webkitGetAsEntry().
+ *
+ * In webkit browsers however, item.webkitGetAsEntry() returns null, because our File()
+ * instances don't actually correspond with a file on disk. But the property is writeable,
+ * so we're able to override it with our own implementation that returns a proper object.
+ */
+const tryMockWebkit = (item) => {
+  try {
+    item.webkitGetAsEntry = () => {
+      return {
+        isFile: true,
+        file: (callback) => callback(item.getAsFile()),
+      }
+    }
+  } catch (e) {
+    // We're in Firefox, this is fine.
+  }
+
+  return item
+}
+
+const createDataTransfer = (files: Cypress.FileReferenceObject[]): DataTransfer => {
   const dataTransfer = new DataTransfer()
 
   files.forEach(({ contents, fileName, lastModified, mimeType }) => {
-    const file = new File([contents], fileName || '', { lastModified, type: mimeType })
+    dataTransfer.items.add(new File([contents], fileName || '', {
+      lastModified,
+      type: mimeType,
+    }))
+  })
+
+  const oldItems = dataTransfer.items
 
-    dataTransfer.items.add(file)
+  // dataTransfer.items is a getter, and it - and the items read from its
+  // returned DataTransferItemList - cannot be assigned to. DataTransferItemLists
+  // also cannot be constructed, so we have to use an array instead.
+  Object.defineProperty(dataTransfer, 'items', {
+    get () {
+      return _.map(oldItems, tryMockWebkit)
+    },
   })
 
-  return dataTransfer.files
+  return dataTransfer
 }
 
 interface InternalAttachFileOptions extends Cypress.AttachFileOptions {
   _log: any
   $el: JQuery
 }
 
+const ACTIONS = {
+  input: (element, dataTransfer, coords, state) => {
+    (element as HTMLInputElement).files = dataTransfer.files
+    const inputEventOptions = addEventCoords({
+      bubbles: true,
+      composed: true,
+    }, coords)
+
+    const changeEventOptions = addEventCoords({
+      bubbles: true,
+    }, coords)
+
+    dispatch(element, state('window'), 'input', inputEventOptions)
+    dispatch(element, state('window'), 'change', changeEventOptions)
+  },
+  'drag-n-drop': (element, dataTransfer, coords, state) => {
```

Comment:
Just curious, will this same logic apply for a drag & drop action in general?

Reply:
I'm not quite sure I follow. What do you mean drag and drop action in general?

Reply:
Yeah, I  was thinking a mouse drag & drop combination but this implementation is specific to uploading files so my question didn't make sense in this context.